src/Functions/FunctionsHashing.h

```diff
@@ -1688,6 +1688,25 @@ struct ImplWyHash64
     static constexpr bool use_int_hash_for_pods = false;
 };
 
+struct ImplEntropyLearnedHash
+{
+    static constexpr auto name = "entropyLearnedHash";
+    // TODO: made also a function, which would set positions by training the method on some dataset
```

Comment:
Yepp, that would be useful 😄 

What about a signature (in SQL) like this: `trainEntropyLearnedHash(data, id)` where `data` is the column with training data and `id` is a string identifier of the dataset. A training run updates a global singleton which maps identifiers to partial keys. And then there will be SQL function `entropyLearnedHash(data, id)` which hashes (usually another) dataset with a partial key identified by `id`.

Ideally, the hash function for `entropyLearnedHash` can also be specified as a third parameter, e.g. `cityHash`, `wyHash` etc. In C++, there could maybe be a dispatch based to different template instantiations where the template parameters are existing hash implementations, e.g. `ImplXXHash64`, `ImplXXH3` etc.

Feel free to implement that outside the existing hash classes. E.g. `FunctionURLHash` could be some inspiration.

Reply:
> What about a signature (in SQL) like this: `trainEntropyLearnedHash(data, id)` where `data` is the column with training data and `id` is a string identifier of the dataset.

Hello, I'm now working on this, I tried to implement a function `trainEntropyLearnedHash(data, id)` like is done in `FunctionURLHash`, I don't understand some things now:
1. `FunctionURLHash` returns a column, in class `FunctionURLHash` it is implemented as the `ColumnPtr`. But `trainEntropyLearnedHash` returns void, so, is it possible to return `void` from the method `executeImpl` of class `FunctionTrainEntropyLearnedHash`?
2. `FunctionURLHash` is placed in `src/Functions/FunctionsHashing.h` file, but `FunctionTrainEntropyLearnedHash` is not a real hash functions, it just changes a some global state. Is it important to place it in this file or not in this one, or does it not matter?

Reply:
1. Good question. All functions are supposed to return "something", and the C++ function interface enforces to return a `ColumnPtr` (there is no `ColumnVoid` class (or sth. like that) and and returning `nullptr` doesn't work as far as I see). Interestingly, function [sleep](https://clickhouse.com/docs/en/sql-reference/functions/other-functions#sleepseconds) has the same "problem". It returns a const column with fixed value (Functions/sleep.h) - this approach has the least runtime overhead compared. Another, more advanced alternative is to return a stringified/serialized representation of the partial key in each row. This has the advantage that one could potentially chain calls like this: `entropyLearnedHash(data, trainEntropyLearnedHash(data))`. That would actually be the ideal solution but for the prototype it isn't necessary.

2. See 1. `getReturnTypeImpl` needs to return either `DataTypeUInt64` (if you like to write integer constant 0 in each row) or `DataTypeString` (for the serialized partial key).

3. Doesn't matter. Let's make it work first, and beautiful second 😄

Reply:
Thank you for the answer. 
Now I'm thinking about the function `entropyLearnedHash(data, id)`. As I see, it also take a column, not just one string to hash. Not this signature is clear for me. But first I thought that the function have to take only one string, as it made with all function from file `src/Functions/FunctionsHashing.h`, for example, for function from struct `ImplWyHash64`. It take only one string like this:
```    static UInt64 apply(const char * s, const size_t len)
    {
        return wyhash(s, len, 0, _wyp);
    }```
I see that it is used in 
```using FunctionWyHash64 = FunctionAnyHash<ImplWyHash64>;```
So, my question is, is it necessary to implement also a function, which takes a single key, not a column of key?

Reply:
I am not sure if I understand your question correctly. The `apply()` method hashes the content of a single row (or more precise: a single cell). It is called once per row from `FunctionAnyHash<ImplWyHash64>`.

In general, `entropyLearnedHash()` will take a column and an id as input. This is similar to [regular hash functions](https://clickhouse.com/docs/en/sql-reference/functions/hash-functions) which only take a column. I see two options:
- (more beautiful option): use `FunctionAnyHash` to implement `entropyLearnedHash()`. For this, you need to change `FunctionAnyHash` such that it understands the `id` column as id and not as input to be hashed (see `FunctionAnyHash::isVariadic()/getReturnTypeImpl()/executeImpl()`.
- Make a new class specifically for `entropyLearnedHash()` which inherits from `IFunction`. There are a bunch of methods to override for that, you could use `FunctionAnyHash` as inspiration.